### ISSUE TITLE
Connection Pooling / Multiplexing in a Protocol Gateway scenario

### ISSUE THREAD

**Issue by Kampfgnom** [Azure/azure-iot-sdk-java#783]
# Description of the issue:

Hi there 👋 

We're implementing a "Protocol Gateway"-like cloud service in Kotlin with Spring. We want devices that are connected to our service to be represented in an upstream Azure IoT Hub seamlessly. We expect to connect a non-trivial amount of devices.

For this reason we're investigating possible optimizations of the connection to IoT Hubs. We are certain that we will need connection pooling and multiplexing in order to support our use cases.

Our devices are not connected to our service at all times and we definitely need to be able to "add and remove" devices from a pooled connection towards IoT Hub.

As the README states, those features (I do not fully know where you're drawing the line) are not yet fully implemented it seems.

# Code sample exhibiting the issue:

The following code sample shows how we currently think we need to use the `TransportClient` in order to do what we want, and how it does not work currently:

```kotlin
val transportClient = TransportClient(IotHubClientProtocol.AMQPS)
 
val gatewayConnStr = "HostName=<HUB>.azure-devices.net;DeviceId=gateway;SharedAccessKey=<GATEWAY_KEY>"
val gatewayClient = DeviceClient(gatewayConnStr, transportClient)
 
val device1Client = DeviceClient(device1ConnStr, transportClient)
val device2Client = DeviceClient(device2ConnStr, transportClient)

transportClient.open()

// use device1Client, maybe one of the devices goes offline in between,
// but we keep the connection open.
 
// Currently it is not allowed to add a new device to an already open connection:
val device3Client = DeviceClient(device3ConnStr, transportClient)
```

Can you help us in our future endeavors? 🙂

**Comment by timtay-microsoft** [comment]
Hi, that is indeed a current limitation on this SDK. We'll take this as a feature request, though.

**Comment by timtay-microsoft** [comment]
We are putting out a preview release with this feature, and we'd love to hear your feedback on it. The relevant commit can be seen here: https://github.com/Azure/azure-iot-sdk-java/commit/fe377ac01d143e3bfa2f00d0433ffcc1f3831f6d

**Comment by timtay-microsoft** [comment]
The preview release in question can be seen here: https://github.com/Azure/azure-iot-sdk-java/releases/tag/2020-11-16-preview

And a sample of this new code can be seen here: https://github.com/Azure/azure-iot-sdk-java/blob/preview/device/iot-device-samples/multiplexing-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/MultiplexingSample.java

**Comment by niklas-wulf-gcx** [comment]
Thanks for the update @timtay-microsoft! OP here, this time from a more official GitHub account ;)

The sample looks exactly like we expected and actually fits our own internal work-around API quite closely 👍 

It should be easy for us to give the preview a test run. Give us a few days to finish other important stuff and we'll report our findings.

**Comment by timtay-microsoft** [comment]
Much appreciated, and thanks for your patience on this. Just in case you need it, here are some helpful links:

The release in question:
https://github.com/Azure/azure-iot-sdk-java/releases/tag/2020-11-16-preview

New client:
https://github.com/Azure/azure-iot-sdk-java/blob/preview/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/MultiplexingClient.java

Sample code:
https://github.com/Azure/azure-iot-sdk-java/blob/preview/device/iot-device-samples/multiplexing-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/MultiplexingSample.java

**Comment by timtay-microsoft** [comment]
@niklas-wulf-gcx in case you missed it, our team has identified a few bugs in the released preview version that may be of interest to you. We'll be putting out a new preview release soon with these fixes. For reference, #1009

**Comment by timtay-microsoft** [comment]
Just a heads up, we've done another formal release of this preview Jar (1.29.0-preview-001) that I recommend picking up for testing on as it has a few good bug fixes in it and some small breaking changes around thrown exceptions. Details can be seen [here](https://github.com/Azure/azure-iot-sdk-java/releases/tag/2020-12-09-preview)

**Comment by timtay-microsoft** [comment]
Just a heads up, we are planning on bringing this preview feature into our normal releases, and the first release with it will be coming out in the next few days. Once that package is released, I'll close this issue

**Comment by timtay-microsoft** [comment]
Closing this issue because we have released device client version 1.29.0 that contains this feature. We won't be pushing new preview releases since the preview feature has successfully migrated to our master releases.

**Comment by az-iot-builder-01** [comment]
@Kampfgnom, @timtay-microsoft, @niklas-wulf-gcx, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
